gluoncv/loss.py

```diff
@@ -160,3 +160,91 @@ def forward(self, cls_pred, box_pred, cls_target, box_target):
             sum_losses.append(cls_losses[-1] + self._lambd * box_losses[-1])
 
         return sum_losses, cls_losses, box_losses
+
+class SoftmaxCrossEntropyLoss(Loss):
+    r"""SoftmaxCrossEntropyLoss with ignore labels
+
+    Parameters
+    ----------
+    axis : int, default -1
+        The axis to sum over when computing softmax and entropy.
+    sparse_label : bool, default True
+        Whether label is an integer array instead of probability distribution.
+    from_logits : bool, default False
+        Whether input is a log probability (usually from log_softmax) instead
+        of unnormalized numbers.
+    weight : float or None
+        Global scalar weight for loss.
+    batch_axis : int, default 0
+        The axis that represents mini-batch.
+    ignore_label : int, default -1
+        The label to ignore.
+    size_average : bool, default False
+        Whether to re-scale loss with regard to ignored labels.
+    valid_size : int, default 1000
+        The size of label, equal to the number of classes for classification models.
+        It affects the re-scaled loss if size_average is True.
+    """
+    # pylint: disable=unused-argument
+    def __init__(self, axis=1, sparse_label=True, from_logits=False, weight=None,
+                 batch_axis=0, ignore_label=-1, size_average=False, valid_size=1000, **kwargs):
+        super(SoftmaxCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
+        self._axis = axis
+        self._sparse_label = sparse_label
+        self._from_logits = from_logits
+        self._ignore_label = ignore_label
+        self._size_average = size_average
+
+    def hybrid_forward(self, F, pred, label, sample_weight=None):
+        """Compute loss"""
+        if not self._from_logits:
+            pred = F.log_softmax(pred, axis=self._axis)
+        if self._sparse_label:
+            if self._size_average:
+                valid_label_map = (label != self._ignore_label).astype('float32')
+            loss = -F.pick(pred, label, axis=self._axis, keepdims=True)
+            loss = F.where(label.expand_dims(axis=self._axis) == self._ignore_label,
+                           F.zeros_like(loss), loss)
+        else:
+            label = _reshape_like(F, label, pred)
+            loss = -F.sum(pred*label, axis=self._axis, keepdims=True)
+        loss = _apply_weighting(F, loss, self._weight, sample_weight)
+        if self._size_average and self._sparse_label:
+            return F.mean(loss, axis=self._batch_axis, exclude=True) * \
+                valid_size / F.sum(valid_label_map)
+        else:
+            return F.mean(loss, axis=self._batch_axis, exclude=True)
+
+class SoftmaxCrossEntropyLossWithAux(SoftmaxCrossEntropyLoss):
```

Comment:
Looks complicated. I don't think this is necessary. How about use two separate losses and add them in the training loop?

Reply:
In segmentation it is also being used often. @zhanghang1989